src/main/java/tc/oc/pgm/restart/RestartListener.java

```diff
@@ -0,0 +1,105 @@
+package tc.oc.pgm.restart;
+
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import org.bukkit.Bukkit;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.plugin.Plugin;
+import org.joda.time.Duration;
+import tc.oc.pgm.Config;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchManager;
+import tc.oc.pgm.api.match.event.MatchFinishEvent;
+import tc.oc.pgm.api.match.event.MatchLoadEvent;
+import tc.oc.pgm.countdowns.SingleCountdownContext;
+import tc.oc.pgm.events.PlayerPartyChangeEvent;
+import tc.oc.util.logging.ClassLogger;
+
+public class RestartListener implements Listener {
+
+  private final Plugin plugin;
+  private final MatchManager matchManager;
+  private final Logger logger;
+
+  private int matchCount;
+  private @Nullable Integer matchLimit;
+  private @Nullable RequestRestartEvent.Deferral deferral;
+
+  public RestartListener(Plugin plugin, MatchManager matchManager) {
+    this.plugin = plugin;
+    this.matchManager = matchManager;
+    this.logger = ClassLogger.get(this.plugin.getLogger(), this.getClass());
+    this.matchLimit = Config.AutoRestart.matchLimit() > 0 ? Config.AutoRestart.matchLimit() : null;
+  }
+
+  private void attemptMatchEnd(Match match) {
+    if (this.deferral == null) return;
+
+    if (match.isRunning()) {
+      if (match.getParticipants().isEmpty()) {
+        this.logger.info("Ending empty match due to restart request");
+        match.finish();
+      }
+    }
+  }
+
+  @EventHandler
+  public void onRequestRestart(RequestRestartEvent event) {
+    if (this.plugin.getServer().getOnlinePlayers().isEmpty()) {
+      Bukkit.getServer().shutdown();
+    } else {
+      Match match = this.matchManager.getMatches().iterator().next();

```

Comment:
Need to check if the match list is not empty, this will throw an exception if no matches are there.